sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,458 @@
+ *
+ * Must be initialized using epmp_transaction_init before use.
+ *
+ * This structure should be considered opaque.
+ */
+typedef struct epmp_transaction {
+  epmp_entry_count_t remaining;
+  epmp_entry_count_t done;
+  epmp_entry_count_t total;  // remaining + done
+  // TODO: add checksum tracker
+} epmp_transaction_t;
+
+/**
+ * EPMP generic status codes.
+ *
+ * These error codes can be used by any function. However if a function
+ * requires additional status codes, it must define a set of status codes to
+ * be used exclusively by such function.
+ */
+typedef enum epmp_result {
+  kEpmpOk = 0, /**< Success. */
+  kEpmpError,  /**< General error. */
+  kEpmpBadArg, /**< Input parameter is not valid. */
+} epmp_result_t;
+
+/**
+ * Start an EPMP transaction.
+ *
+ * The transaction will be initialized to expect the given number of entries
+ * to be reconfigured. Only one transaction may be in progress at a time.
+ * Call epmp_transaction_push to finalize the transaction and update the
+ * EPMP control registers.
+ *
+ * @param transaction Transaction to initialize.
+ * @param count Number of regions to be configured.
+ * @returns `epmp_result_t` // TODO: more specialized return code?
+ */
+epmp_result_t epmp_transaction_init(epmp_transaction_t *transaction,
+                                    epmp_entry_count_t count);
+
+/**
+ * Finish an EPMP transaction and update EPMP control registers.
+ *
+ * Check that the expected number of regions have been reconfigured and
+ * then update the EPMP control registers. Once this call returns the
+ * transaction will be complete and a new transaction should be
+ * initialized if further updates are required.
+ *
+ * Updates will occur in the following sequence:
+ *
+ *   1. pmpaddr0-pmpaddr15
+ *   2. pmpcfg0-pmpcfg3
+ *
+ * TODO: should there be a way to abandon errored transactions?
+ *
+ * @param transaction Transaction to finalize.
+ * @param count Number of regions that should have been configured.
+ * @returns `epmp_result_t` // TODO: more specialized return code?
+ */
+epmp_result_t epmp_transaction_push(epmp_transaction_t *transaction,
+                                    epmp_entry_count_t count);
+
+/**
+ * EPMP configuration status codes.
+ */
+typedef enum epmp_entry_configure_result {
+  kEpmpRegionConfigureOk = kEpmpOk,
+  kEpmpRegionConfigureError = kEpmpError,
+  kEpmpRegionConfigureBadArg = kEpmpBadArg,
+
+  /**
+   * Invalid addresses provided for the selected encoding method.
+   *
+   * TODO: is it useful to split this further? (e.g. alignment, size)
+   */
+  kEpmpRegionConfigureBadRegion,
+
+  /**
+   * Encoding the entry would interfere with a different pre-existing entry.
+   *
+   * New entries will be rejected if they:
+   *  * Modify the base address of a pre-existing TOR entry.
+   *  * Would result in an address being used in both a NAPOT/NA4 entry and a
+   *    TOR entry.
+   *
+   * TODO: this means that TOR stacks need to be disabled top to bottom and then
```

Comment:
By stack do you mean a single TOR region `pmpcfg` and `pmpaddr` pair?

Reply:
What I mean by 'stacking' is arranging TOR regions so that they are adjacent and therefore take only N+1 entry slots. For example:

```
entry | pmpaddr | pmpcfg
0     | 0x10    | OFF
1     | 0x24    | TOR (start = 0x10, end = 0x24)
2     | 0x50    | TOR (start = 0x24, end = 0x50)
3     | 0x78    | TOR (start = 0x50, end = 0x78)
```

I'll add that to the description.

Reply:
What this comment is trying to say is that to move the boundary between entries 1 and 2 (say from `0x24` to `0x28`) you would first have to disable entry 2, modify entry 1 and then re-enable entry 2.

It looks like the existing plan is to have static regions so this probably wouldn't occur in practice.